File: clients/haskell/src/Model/Player.hs
module Model.Player where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int
import Model.Vehicle (Vehicle)

-- | Player (game participant)
data Player = Player {
    -- | Index
    index :: Int32,
    -- | Current score
    score :: Int64,
    -- | List of player's vehicles
    vehicles :: [Vehicle] }
    deriving Show

instance Trans Player where
    read = do
        index <- Trans.read
        score <- Trans.read
        vehicles <- Trans.read
        return Player {
            index,
            score,
            vehicles }
    
    write Player {
        index,
        score,
        vehicles } = do
            Trans.write index
            Trans.write score
            Trans.write vehicles

File: clients/rust/src/model/player.rs
use super::*;

/// Player (game participant)
#[derive(Clone, Debug)]
pub struct Player {
    /// Index
    pub index: i32,
    /// Current score
    pub score: i64,
    /// List of player's vehicles
    pub vehicles: Vec<model::Vehicle>,
}

impl trans::Trans for Player {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.index.write_to(writer)?;
        self.score.write_to(writer)?;
        self.vehicles.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let index: i32 = trans::Trans::read_from(reader)?;
        let score: i64 = trans::Trans::read_from(reader)?;
        let vehicles: Vec<model::Vehicle> = trans::Trans::read_from(reader)?;
        Ok(Self {
            index,
            score,
            vehicles,
        })
    }
}